Modify Regex Dependency in Validator Class

hsh108868 opened the pull request:
관련 이슈: [Validator 클라스 안의 Regex 의존성 수정하기 #70 ](https://github.com/devrel-kr/nhn-toast-notification-service-custom-connector/issues/70)
작업 사항: nt-sms 내부 Validator 클래스에 있는 Regex 객체들을 싱글톤으로 구조 변경하여 개별 클래스와 인터페이스로 분리하고 해당 Regex 객체 바탕으로 Validator의 Regex 의존성 수정

hsh108868 committed:
singleton

hsh108868 committed:
add singleton for regex

hsh108868 committed:
Merge branch 'main' of https://github.com/hsh108868/nhn-toast-notification-service-custom-connector

hsh108868 committed:
Merge branch 'devrel-kr:main' into main

oy-hsNam committed:
[feature]return을 바꿔야 하지 않을까 해서 함.

hsh108868 committed:
add parameter in validator class

oy-hsNam committed:
feature 선언부 옮김

hsh108868 committed:
테스트 코드 수정

hsh108868 committed:
Merge branch 'devrel-kr:main' into main

MultiDCP committed:
Move Wrapper to own File

Move RegexDateTimeWrapper class and interface to their own file for other codes using them

MultiDCP committed:
Replace IsValidDateFormat Class with Regex Wrapper

Replace IsValidDateFormat Class in SendMessagesRequestBodyValidator with Regex Wrapper in RegexDateTimeWrapper Class, and Update Test codes

MultiDCP committed:
Replace IsValidDateFormat Class

Replace IsValidDateFormat Class in ListMessageStatusQueryValidator with Regex Wrapper in RegexDateTimeWrapper Class, and Update Test codes

hsh108868 committed:
Merge pull request #1 from MultiDCP/main

Regex Wrapper 코드 별도 분리 및 타 Validator 구조 수정

justinyoo commented on src/nt-sms/Validators/ListMessageStatusQueryValidator.cs:
```diff
@@ -45,10 +45,10 @@ public class ListMessageStatusRequestQueryValidator : AbstractValidator<ListMess
         /// Initializes a new instance of the <see cref="ListMessageStatusRequestQueryValidator"/> class.
         /// </summary>
-        public ListMessageStatusRequestQueryValidator()
+        public ListMessageStatusRequestQueryValidator(IRegexDateTimeWrapper iRegexDateTimeWrapper)
```
이건 마이너한 내용인데, 파라미터 이름을 `regex` 정도로 줄여도 상관 없습니다. 만약에 둘 이상의 regex 인스턴스가 들어온다면 그 땐 `regexDateTime` 이라든가 하는 식으로 하는게 낫겠지만요.

justinyoo commented on src/nt-sms/Validators/ListMessageStatusQueryValidator.cs:
```diff
@@ -58,17 +58,5 @@ public ListMessageStatusRequestQueryValidator()
-            {
-                return Regex.IsMatch(date, @"^([0-9][0-9][0-9][0-9])-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[0-1]) ([01][0-9]|2[0123]):([0-5][0-9]):([0-5][0-9])$");
-            }   
-        }
```
요 메소드는 살려두시고, 이 안에서 `return this._regex.IsMatch(date);` 를 호출하면 좋습니다.

justinyoo commented on src/nt-sms/Validators/ListMessageStatusQueryValidator.cs:
```diff
@@ -45,10 +45,10 @@ public class ListMessageStatusRequestQueryValidator : AbstractValidator<ListMess
-            this.RuleFor(p => p.StartUpdateDate).Must(IsValidDateFormat).NotEmpty();
-            this.RuleFor(p => p.EndUpdateDate).Must(IsValidDateFormat).NotEmpty().GreaterThan(q => q.StartUpdateDate);
+            this.RuleFor(p => p.StartUpdateDate).Must(iRegexDateTimeWrapper.IsMatch).NotEmpty();
+            this.RuleFor(p => p.EndUpdateDate).Must(iRegexDateTimeWrapper.IsMatch).NotEmpty().GreaterThan(q => q.StartUpdateDate);
```
그러면 여기는 건드릴 필요가 없겠죠?

justinyoo commented on src/nt-sms/Validators/ListMessageStatusQueryValidator.cs:
```diff
@@ -45,10 +45,10 @@ public class ListMessageStatusRequestQueryValidator : AbstractValidator<ListMess
         /// <summary>
```
단, 여기에 `private readonly IRegexDateTimeWrapper _regex;` 를 프라이빗 필드로 정의하고, 생성자 안에서 파라미터를 이 프라이빗 필드로 할당하는 작업을 맨 처음에 추가해 놓으면 됩니다.

justinyoo commented on src/nt-sms/Validators/ListMessagesQueryValidator.cs:
```diff
@@ -36,12 +36,13 @@ public static async Task<ListMessagesRequestQueries> Validate(this Task<ListMess
         /// Initializes a new instance of the <see cref="ListMessagesRequestQueryValidator"/> class.
         /// </summary>
-        public ListMessagesRequestQueryValidator()
+        public ListMessagesRequestQueryValidator(IRegexDateTimeWrapper iRegexDateTimeWrapper)
```
여기도 마찬가지

justinyoo commented on src/nt-sms/Validators/ListMessagesQueryValidator.cs:
```diff
@@ -52,39 +53,40 @@ public ListMessagesRequestQueryValidator()
             this.RuleFor(p => p.StartRequestDate)
                 .NotEmpty()
-                .Must(IsValidDateFormat)
+                .Must(iRegexDateTimeWrapper.IsMatch)
```
여기도 마찬가지

justinyoo commented on src/nt-sms/Validators/ListMessagesQueryValidator.cs:
```diff
@@ -52,39 +53,40 @@ public ListMessagesRequestQueryValidator()
             this.RuleFor(p => p.EndRequestDate)
                 .NotEmpty()
-                .Must(IsValidDateFormat)
+                .Must(iRegexDateTimeWrapper.IsMatch)
```
여기도 마찬가지

justinyoo commented on src/nt-sms/Validators/ListMessagesQueryValidator.cs:
```diff
@@ -52,39 +53,40 @@ public ListMessagesRequestQueryValidator()
             this.RuleFor(p => p.StartCreateDate)
                 .NotEmpty()
-                .Must(IsValidDateFormat)
+                .Must(iRegexDateTimeWrapper.IsMatch)
```
여기도 마찬가지

justinyoo commented on src/nt-sms/Validators/ListMessagesQueryValidator.cs:
```diff
@@ -52,39 +53,40 @@ public ListMessagesRequestQueryValidator()
             this.RuleFor(p => p.EndCreateDate)
                 .NotEmpty()
-                .Must(IsValidDateFormat)
+                .Must(iRegexDateTimeWrapper.IsMatch)
```
여기도 마찬가지

justinyoo commented on src/nt-sms/Validators/ListMessagesQueryValidator.cs:
```diff
@@ -52,39 +53,40 @@ public ListMessagesRequestQueryValidator()
+            this.RuleFor(p => p.StartResultDate).Must(iRegexDateTimeWrapper.IsMatch).When(p => p.StartResultDate != null);
 
-            this.RuleFor(p => p.EndResultDate).Must(IsValidDateFormat).GreaterThan(p => p.StartResultDate).When(p => p.EndResultDate != null);
+            this.RuleFor(p => p.EndResultDate).Must(iRegexDateTimeWrapper.IsMatch).GreaterThan(p => p.StartResultDate).When(p => p.EndResultDate != null);
```
여기도 마찬가지

justinyoo commented on src/nt-sms/Validators/ListMessagesQueryValidator.cs:
```diff
@@ -110,16 +112,6 @@ public ListMessagesRequestQueryValidator()
-            {
-                return Regex.IsMatch(date, @"^([0-9][0-9][0-9][0-9])-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[0-1]) ([01][0-9]|2[0123]):([0-5][0-9]):([0-5][0-9])$");
-            }
-        }
```
여기도 마찬가지

justinyoo commented on src/nt-sms/Validators/RegexDateTimeWrapper.cs:
```diff
@@ -0,0 +1,29 @@
+{
+    /// <summary>
+    /// This represents the singleton entity of regex for the validators.
+    /// </summary>
```
요기 코멘트 내용은 클래스에 붙는 거니까 확인해 보시구요. 인터페이스용 코멘트는 다른 인터페이스 참조해서 다시 달아주세욥.

justinyoo commented on src/nt-sms/Validators/RegexDateTimeWrapper.cs:
```diff
@@ -0,0 +1,29 @@
+    /// </summary>
+    public interface IRegexDateTimeWrapper
+    {
+        bool IsMatch(string date);
```
여기에는 코멘트 붙여 주시구요. 다른 `bool` 반환하는 메소드 찾아서 확인해 보시고 비슷한 패턴으로 달아주세욥.

justinyoo commented on src/nt-sms/Validators/RegexDateTimeWrapper.cs:
```diff
@@ -0,0 +1,29 @@
+        bool IsMatch(string date);
+    }
+
+    public class RegexDateTimeWrapper : IRegexDateTimeWrapper
```
여기에는 클래스 코멘트 달아주시구요

justinyoo commented on src/nt-sms/Validators/RegexDateTimeWrapper.cs:
```diff
@@ -0,0 +1,29 @@
+    {
+        private static Regex _regex = new Regex(@"^([0-9][0-9][0-9][0-9])-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[0-1]) ([01][0-9]|2[0123]):([0-5][0-9]):([0-5][0-9])$");
+
+        public bool IsMatch(string date)
```
여기에 붙는 코멘트는 이미 인터페이스에 달아뒀으니 `/// <inheridtoc />` 정도로만 달아두셔도 됩니다.

justinyoo commented on src/nt-sms/Validators/SendMessagesRequestBodyValidator.cs:
```diff
@@ -42,22 +42,17 @@ public class SendMessagesRequestBodyValidator : AbstractValidator<SendMessagesRe
         /// </summary>
         ///
-        public SendMessagesRequestBodyValidator()
+        public SendMessagesRequestBodyValidator(IRegexDateTimeWrapper iRegexDateTimeWrapper)
```
앞서와 마찬가지로 해 주세요

justinyoo commented on src/nt-sms/Validators/SendMessagesRequestBodyValidator.cs:
```diff
@@ -42,22 +42,17 @@ public class SendMessagesRequestBodyValidator : AbstractValidator<SendMessagesRe
             this.RuleFor(p => p.Body).NotNull().MaximumLength(255);
             this.RuleFor(p => p.SenderNumber).NotEmpty().MaximumLength(13);
-            this.RuleFor(p => p.RequestDate).Must(IsValidDateFormat).When(p => !string.IsNullOrWhiteSpace(p.RequestDate));
+            this.RuleFor(p => p.RequestDate).Must(iRegexDateTimeWrapper.IsMatch).When(p => !string.IsNullOrWhiteSpace(p.RequestDate));
```
여기도 마찬가지

justinyoo commented on src/nt-sms/Validators/SendMessagesRequestBodyValidator.cs:
```diff
@@ -42,22 +42,17 @@ public class SendMessagesRequestBodyValidator : AbstractValidator<SendMessagesRe
-        private bool IsValidDateFormat(string date)
-        {
-            return Regex.IsMatch(date, @"^([0-9][0-9][0-9][0-9])-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[0-1]) ([01][0-9]|2[0123]):([0-5][0-9]):([0-5][0-9])$");
-        }
```
여기도 마찬가지

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessageStatusRequestQueryValidatorTests.cs:
```diff
@@ -35,7 +35,9 @@ public void Given_Values_When_Validate_Invoked_Then_It_Should_Return_Result(stri
             };
-            var validator = new ListMessageStatusRequestQueryValidator();
+            
+            IRegexDateTimeWrapper iRegexDateTimeWrapper = new RegexDateTimeWrapper();
```
이 메소드는 `ListMessageStatusRequestQueryValidator.Validate()` 메소드를 테스트하는 테스트 메소드예요. 따라서, 우리가 인터페이스를 만든 목적에 맞게 `var wrapper = new Mock<IRegexDateTimeWrapper>();` 정도로 목킹하시고, 그 다음에서 `wrapper.Setup(p => p.IsMatch(It.IsAny<string>()).Returns(true);` 또는 `wrapper.Setup(p => p.IsMatch(It.IsAny<string>()).Returns(false);` 정도로 결과값을 "했다치고" 형태로 만들어서 테스트 하십쇼.

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessageStatusRequestQueryValidatorTests.cs:
```diff
@@ -62,7 +64,9 @@ public void Given_InvalidValues_When_Validate_Invoked_Then_It_Should_Throw_Excep
             };
-            var validator = new ListMessageStatusRequestQueryValidator();
+
+            IRegexDateTimeWrapper iRegexDateTimeWrapper = new RegexDateTimeWrapper();
```
여기도 마찬가지

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessageStatusRequestQueryValidatorTests.cs:
```diff
@@ -84,7 +88,9 @@ public async Task Given_ValidValues_When_Validate_Invoked_Then_It_Should_Return_
             };
-            var validator = new ListMessageStatusRequestQueryValidator();
+            
+            IRegexDateTimeWrapper iRegexDateTimeWrapper = new RegexDateTimeWrapper();
```
여기도 마찬가지

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessagesRequestQueryValidatorTests.cs:
```diff
@@ -56,7 +56,9 @@ public class ListMessagesRequestQueryValidatorTests
             };
-            var validator = new ListMessagesRequestQueryValidator();
+
+            IRegexDateTimeWrapper iRegexDateTimeWrapper = new RegexDateTimeWrapper();
```
여기도 마찬가지

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessagesRequestQueryValidatorTests.cs:
```diff
@@ -105,7 +107,9 @@ public class ListMessagesRequestQueryValidatorTests
             };
-            var validator = new ListMessagesRequestQueryValidator();
+
+            IRegexDateTimeWrapper iRegexDateTimeWrapper = new RegexDateTimeWrapper();
```
여기도 마찬가지

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessagesRequestQueryValidatorTests.cs:
```diff
@@ -141,7 +145,9 @@ public class ListMessagesRequestQueryValidatorTests
             };
-            var validator = new ListMessagesRequestQueryValidator();
+
+            IRegexDateTimeWrapper iRegexDateTimeWrapper = new RegexDateTimeWrapper();
```
여기도 마찬가지

justinyoo commented on test/NhnToast.Sms.Tests/Validators/SendMessagesRequestBodyValidatorTests.cs:
```diff
@@ -48,7 +48,9 @@ public class SendMessagesRequestBodyValidatorTests
             };
-            var validator = new SendMessagesRequestBodyValidator();
+
+            IRegexDateTimeWrapper iRegexDateTimeWrapper = new RegexDateTimeWrapper();
```
여기도 마찬가지

justinyoo commented on test/NhnToast.Sms.Tests/Validators/SendMessagesRequestBodyValidatorTests.cs:
```diff
@@ -88,7 +90,9 @@ public class SendMessagesRequestBodyValidatorTests
             };
-            var validator = new SendMessagesRequestBodyValidator();
+
+            IRegexDateTimeWrapper iRegexDateTimeWrapper = new RegexDateTimeWrapper();
```
여기도 마찬가지

justinyoo commented on test/NhnToast.Sms.Tests/Validators/SendMessagesRequestBodyValidatorTests.cs:
```diff
@@ -126,7 +130,9 @@ public class SendMessagesRequestBodyValidatorTests
             };
-            var validator = new SendMessagesRequestBodyValidator();
+
+            IRegexDateTimeWrapper iRegexDateTimeWrapper = new RegexDateTimeWrapper();
```
여기도 마찬가지

justinyoo reviewed (changes requested):
리뷰 남겨뒀습니다! 거의 다 잘 하셨쎄요. 조금만 더 보시면 될 것 같구요, 다만 `RegexDateTimeWrapper` 클래스 자체에 대한 테스트가 없으니 그걸 추가해 주세욥!

MultiDCP committed:
Modify ListMessageStatusQueryValidator

Modify ListMessageStatusQueryValidator as the comment reviewed

MultiDCP committed:
Modify ListMessagesQueryValidator

Modify ListMessagesQueryValidator as the comment reviewed

MultiDCP committed:
Modify SendMessagesRequestBodyValidator

Modify SendMessageRequestBodyValidator as the comment reviewed

MultiDCP committed:
Add Comments for RegexDateTimeWrapper

Add comments for interface, class, and methods in RegexDateTimeWrapper

hsh108868 committed:
Merge pull request #2 from MultiDCP/main

수정사항 반영

justinyoo commented on src/nt-sms/Validators/ListMessageStatusQueryValidator.cs:
```diff
@@ -42,11 +42,14 @@ public static async Task<ListMessageStatusRequestQueries> Validate(this Task<Lis
-        public ListMessageStatusRequestQueryValidator()
+        public ListMessageStatusRequestQueryValidator(IRegexDateTimeWrapper regex)
         {
+            this._regex = regex;
```
`regex` 라는 인스턴스가 들어올 때 null 체크가 빠졌습니다. 다른 생성자 참조해서 null 체크 로직 추가해 주세요.

justinyoo commented on src/nt-sms/Validators/ListMessagesQueryValidator.cs:
```diff
@@ -36,13 +36,16 @@ public static async Task<ListMessagesRequestQueries> Validate(this Task<ListMess
-        public ListMessagesRequestQueryValidator()
+        public ListMessagesRequestQueryValidator(IRegexDateTimeWrapper regex)
         {
+            this._regex = regex;
```
여기도 널 체크

justinyoo commented on src/nt-sms/Validators/ListMessagesQueryValidator.cs:
```diff
@@ -118,8 +119,13 @@ private bool IsValidDateFormat(string date)
             }
         }
+
+        List<string> MsgStatusType = new List<string>() { "0", "1", "2", "3", "4", "5" };
+        List<string> resultCodeType = new List<string>() { "MTR1", "MTR2" };
+        List<string> subResultCodeType = new List<string>() { "MTR2_1", "MTR2_2", "MTR2_3" };
```
요 부분은 나중에 다시 리팩토링한다고 생각해야 할 것 같아요. 지금 PR하고는 상관 없으니 나중에 다시 한 번 보시지욥.

justinyoo commented on src/nt-sms/Validators/SendMessagesRequestBodyValidator.cs:
```diff
@@ -38,12 +38,15 @@ public static async Task<SendMessagesRequestBody> Validate(this Task<SendMessage
-        public SendMessagesRequestBodyValidator()
+        public SendMessagesRequestBodyValidator(IRegexDateTimeWrapper regex)
         {
+            this._regex = regex;
```
여기도 널 체크

justinyoo commented on src/nt-sms/Validators/ListMessageStatusQueryValidator.cs:
```diff
@@ -67,8 +69,10 @@ private bool IsValidDateFormat(string date)
             {
-                return Regex.IsMatch(date, @"^([0-9][0-9][0-9][0-9])-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[0-1]) ([01][0-9]|2[0123]):([0-5][0-9]):([0-5][0-9])$");
-            }   
+                return this._regex.IsMatch(date);
+            }
```
여기서는 굳이 `if (date == null) { ... }` 블록이 필요할까 한 번 고민해 보시면 좋겠습니다.

justinyoo commented on src/nt-sms/Validators/ListMessagesQueryValidator.cs:
```diff
@@ -118,8 +119,13 @@ private bool IsValidDateFormat(string date)
             else
             {
-                return Regex.IsMatch(date, @"^([0-9][0-9][0-9][0-9])-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[0-1]) ([01][0-9]|2[0123]):([0-5][0-9]):([0-5][0-9])$");
+                return this._regex.IsMatch(date);
             }
```
여기서도 굳이 `if (date == null) { ... }` 블록이 필요할까 한 번 고민해 보시면 좋겠습니다.

justinyoo commented on src/nt-sms/Validators/SendMessagesRequestBodyValidator.cs:
```diff
@@ -56,7 +59,14 @@ public SendMessagesRequestBodyValidator()
         private bool IsValidDateFormat(string date)
         {
-            return Regex.IsMatch(date, @"^([0-9][0-9][0-9][0-9])-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[0-1]) ([01][0-9]|2[0123]):([0-5][0-9]):([0-5][0-9])$");
+            if (date == null)
+            {
+                return false;
+            }
+            else
+            {
+                return this._regex.IsMatch(date);
+            }
```
여기서도 굳이 `if (date == null) { ... }` 블록이 필요할까 한 번 고민해 보시면 좋겠습니다.

justinyoo reviewed (changes requested):
곧 `RegexDateTimeWrapper` 클라스를 테스트하는 테스트케이스도 추가해 주신다고 알고 있겠습니다!

oy-hsNam committed:
[Add Test file]RegexDateTime class test file

oy-hsNam committed:
Merge branch 'main' of https://github.com/HsnGoodDev/nhn-toast-notification-service-custom-connector

oy-hsNam committed:
[add test case] 과거와 미래 추가

hsh108868 committed:
Merge pull request #3 from HsnGoodDev/main

[Add Test] test file 추가 합니다.

oy-hsNam committed:
[fix] null check one line

justinyoo commented on test/NhnToast.Sms.Tests/Validators/RegexDateTimeWrapperTests.cs:
```diff
@@ -0,0 +1,30 @@
+        [DataRow("2022-0&-OO 00:oo:OT", false)]
+        public void Given_Values_When_Validate_Invoked_Then_It_Should_Return_Result(string date, bool expected)
+        {
+            IRegexDateTimeWrapper iRegexDateTimeWrapper = new RegexDateTimeWrapper();
```
C#의 컨벤션은 정말로 필요하지 않는 이상 `var`를 사용합니다. 따라서 이 라인은 대략...

```csharp
var wrapper = new RegexDateTimeWrapper();
```

정도로 표현하면 좋습니다.

justinyoo commented on test/NhnToast.Sms.Tests/Validators/RegexDateTimeWrapperTests.cs:
```diff
@@ -0,0 +1,30 @@
+        {
+            IRegexDateTimeWrapper iRegexDateTimeWrapper = new RegexDateTimeWrapper();
+
+            var result = iRegexDateTimeWrapper.IsMatch(date).Should().Be(expected);
```
테스트는 기본적으로 1) Arrange, 2) Act, 3) Assert 의 순서를 따라갑니다. 위의 라인에서 Arrange는 표현했으나, 이 라인에서 Act와 Assert를 한번에 구현했기 때문에 유지보수를 하려는 미래의 나에게는 코드 파악이 어려울 수 있습니다.

따라서, 이 부분은 보통 이런 식으로 표현하면 좋습니다.

```csharp
// Act
var result = wrapper.IsMatch(date);

// Assert
result.Should().Be(expected);
```

oy-hsNam committed:
[feature]test case

oy-hsNam committed:
[feature]add bracket

hsh108868 committed:
Merge pull request #4 from HsnGoodDev/main

[fix] null check one line

oy-hsNam committed:
[feature]삼항연산자 교체

hsh108868 committed:
Merge pull request #5 from HsnGoodDev/main

hsh108868 committed:
add mocking test

justinyoo commented on src/nt-sms/Validators/ListMessageStatusQueryValidator.cs:
```diff
@@ -57,18 +61,12 @@ public ListMessageStatusRequestQueryValidator()
-            {
-                return Regex.IsMatch(date, @"^([0-9][0-9][0-9][0-9])-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[0-1]) ([01][0-9]|2[0123]):([0-5][0-9]):([0-5][0-9])$");
-            }   
+            return (date != null) ? _regex.IsMatch(date) : false ;
```
이미 `Regex.IsMatch(...)` 에서 date 값이 null 인지 아닌지를 체크하는데도 여기서 한 번 더 체크하는 이유가 있나요?

justinyoo commented on src/nt-sms/Validators/ListMessagesQueryValidator.cs:
```diff
@@ -107,19 +111,14 @@ public ListMessagesRequestQueryValidator()
-            {
-                return Regex.IsMatch(date, @"^([0-9][0-9][0-9][0-9])-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[0-1]) ([01][0-9]|2[0123]):([0-5][0-9]):([0-5][0-9])$");
-            }
+            return (date != null) ? _regex.IsMatch(date) : false ;
```
마찬가지로, 이미 `Regex.IsMatch(...)` 에서 date 값이 null 인지 아닌지를 체크하는데도 여기서 한 번 더 체크하는 이유가 있나요?

justinyoo commented on src/nt-sms/Validators/RegexDateTimeWrapper.cs:
```diff
@@ -0,0 +1,29 @@
+        /// <inheritdoc />
+        public bool IsMatch(string date)
+        {
+            return (date != null) ? _regex.IsMatch(date) : false ;
```
여기선 원래대로 if 블록으로 풀어 쓰는 게 가독성이 높습니다.

MultiDCP replied:
RegexDateTimeWrapper 클래스의 IsMatch 함수도 이름을 Regex의 IsMatch 함수와 이름을 달리하는 것이 깔끔할까요? 아니면 당장은 딱히 고치지 않아도 상관없는 부분일까요?

MultiDCP replied:
> 여기선 원래대로 if 블록으로 풀어 쓰는 게 가독성이 높습니다.

RegexDateTimeWrapper 클래스의 IsMatch 함수도 이름을 Regex의 IsMatch 함수와 이름을 달리하는 것이 깔끔할까요? 아니면 당장은 딱히 고치지 않아도 상관없는 부분일까요?

justinyoo replied:
이건 래퍼 클라스이기 때문에 가급적이면 원래 Regex 클래스의 메소드명을 그대로 가져가는 것이 좋습니다.

justinyoo commented on src/nt-sms/Validators/RegexDateTimeWrapper.cs:
```diff
@@ -0,0 +1,29 @@
+    /// </summary>
+    public class RegexDateTimeWrapper : IRegexDateTimeWrapper
+    {
+        private static Regex _regex = new Regex(@"^([0-9][0-9][0-9][0-9])-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[0-1]) ([01][0-9]|2[0123]):([0-5][0-9]):([0-5][0-9])$");
```
정적 필드의 경우에는 private 일 경우 앞에 관례적으로 언더스코어를 붙이지 않습니다.

justinyoo commented on src/nt-sms/Validators/SendMessagesRequestBodyValidator.cs:
```diff
@@ -56,7 +59,7 @@ public SendMessagesRequestBodyValidator()
         private bool IsValidDateFormat(string date)
         {
-            return Regex.IsMatch(date, @"^([0-9][0-9][0-9][0-9])-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[0-1]) ([01][0-9]|2[0123]):([0-5][0-9]):([0-5][0-9])$");
+            return (date != null) ? _regex.IsMatch(date) : false ;
```
마찬가지로, 이미 `Regex.IsMatch(...)` 에서 date 값이 null 인지 아닌지를 체크하는데도 여기서 한 번 더 체크하는 이유가 있나요?

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessageStatusRequestQueryValidatorTests.cs:
```diff
@@ -35,7 +36,10 @@ public void Given_Values_When_Validate_Invoked_Then_It_Should_Return_Result(stri
-            var validator = new ListMessageStatusRequestQueryValidator();
+
+            var wrapper = new Mock<IRegexDateTimeWrapper>();
+            wrapper.Setup(p => p.IsMatch(It.IsAny<string>())).Returns(false);
```
이 값을 `false`로 지정하면 유효성 검사를 통과하지 못했다고 가정하는 건데, 이 테스트가 그게 맞나요?

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessageStatusRequestQueryValidatorTests.cs:
```diff
@@ -62,7 +66,11 @@ public void Given_InvalidValues_When_Validate_Invoked_Then_It_Should_Throw_Excep
-            var validator = new ListMessageStatusRequestQueryValidator();
+
+            var wrapper = new Mock<IRegexDateTimeWrapper>();
+            wrapper.Setup(p => p.IsMatch(It.IsAny<string>())).Returns(true);
```
테스트 메소드 이름에서 이미 `It_Should_Throw_Exception`라고 언급하고 있으니, 이 테스트 메소드는 유효성 검사를 통과하지 못했다는 걸 가정하는 거니까 여기서는 `false` 값이 되어야 하지 않을까요?

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessagesRequestQueryValidatorTests.cs:
```diff
@@ -105,7 +110,10 @@ public class ListMessagesRequestQueryValidatorTests
-            var validator = new ListMessagesRequestQueryValidator();
+
+            var wrapper = new Mock<IRegexDateTimeWrapper>();
+            wrapper.Setup(p => p.IsMatch(It.IsAny<string>())).Returns(true);
```
테스트 메소드 제목을 보시면 여기서 `true` 라고 해야 할 지, `false` 라고 해야 할 지 알 수 있을 겁니다.

justinyoo commented on test/NhnToast.Sms.Tests/Validators/SendMessagesRequestBodyValidatorTests.cs:
```diff
@@ -88,7 +92,10 @@ public class SendMessagesRequestBodyValidatorTests
-            var validator = new SendMessagesRequestBodyValidator();
+
+            var wrapper = new Mock<IRegexDateTimeWrapper>();
+            wrapper.Setup(p => p.IsMatch(It.IsAny<string>())).Returns(true);
```
테스트 메소드 이름을 다시 한 번 살펴보세요

justinyoo reviewed (changes requested):
코멘트 남겨뒀습니다. 테스트가 실패하는 경우를 보니 뭔가 목킹 값을 조정해야 할 것 같습니다.

MultiDCP committed:
Remove Validator Multiple NullCheck

Remove and edit validators' multiple null check structures (now check one time)

MultiDCP committed:
Redo If Block In Wrapper

Redo if-else block in wrapper class

hsh108868 committed:
Modify wrapper setup

MultiDCP committed:
Remove Validator Multiple NullCheck

Remove and edit ListMessageStatusQueryValidators' multiple null check structures

hsh108868 committed:
Merge pull request #6 from MultiDCP/main

justinyoo commented on src/nt-sms/Validators/ListMessageStatusQueryValidator.cs:
```diff
@@ -57,18 +61,12 @@ public ListMessageStatusRequestQueryValidator()
-            {
-                return Regex.IsMatch(date, @"^([0-9][0-9][0-9][0-9])-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[0-1]) ([01][0-9]|2[0123]):([0-5][0-9]):([0-5][0-9])$");
-            }   
+            return _regex.IsMatch(date);
```
52번 라인과 마찬가지로 `_regex` 필드는  메소드 안의 로컬 변수가 아니기 때문에 컨벤션상 `this._regex`로 표현하는 것이 명확합니다.

justinyoo commented on src/nt-sms/Validators/ListMessagesQueryValidator.cs:
```diff
@@ -107,19 +111,14 @@ public ListMessagesRequestQueryValidator()
-            {
-                return Regex.IsMatch(date, @"^([0-9][0-9][0-9][0-9])-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[0-1]) ([01][0-9]|2[0123]):([0-5][0-9]):([0-5][0-9])$");
-            }
+            return _regex.IsMatch(date);
```
48번 라인과 마찬가지로 _regex 필드는 메소드 안의 로컬 변수가 아니기 때문에 컨벤션상 this._regex로 표현하는 것이 명확합니다.

justinyoo commented on src/nt-sms/Validators/RegexDateTimeWrapper.cs:
```diff
@@ -0,0 +1,36 @@
+            {
+                return false;
+            }
+            else
+            {
+                return regex.IsMatch(date);
+            }
```
이미 `if { ... }` 블록 안에서 `return` 구문을 사용했으므로 여기서는 굳이 `else { ... }` 블록이 필요 없습니다. 곧바로 `return` 구문 쓰는 것이 코드 복잡도를 줄이는 방향입니다.

justinyoo commented on src/nt-sms/Validators/SendMessagesRequestBodyValidator.cs:
```diff
@@ -56,7 +59,7 @@ public SendMessagesRequestBodyValidator()
         private bool IsValidDateFormat(string date)
         {
-            return Regex.IsMatch(date, @"^([0-9][0-9][0-9][0-9])-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[0-1]) ([01][0-9]|2[0123]):([0-5][0-9]):([0-5][0-9])$");
+            return _regex.IsMatch(date);
```
48번 라인과 마찬가지로 _regex 필드는 메소드 안의 로컬 변수가 아니기 때문에 컨벤션상 this._regex로 표현하는 것이 명확합니다.

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessageStatusRequestQueryValidatorTests.cs:
```diff
@@ -35,7 +36,10 @@ public void Given_Values_When_Validate_Invoked_Then_It_Should_Return_Result(stri
-            var validator = new ListMessageStatusRequestQueryValidator();
+
+            var wrapper = new Mock<IRegexDateTimeWrapper>();
+            wrapper.Setup(p => p.IsMatch(It.IsAny<string>())).Returns(expected);
```
여기서 `expected` 값은 이 `Validator.Validate(...)` 메소드의 결과값에 대한 예상치이지 이 목킹 결과값에 대한 예상치가 아니므로 이렇게 작성하면 테스트를 무조건 통과하는 의미없는 테스트 코드가 됩니다.

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessagesRequestQueryValidatorTests.cs:
```diff
@@ -56,7 +58,10 @@ public class ListMessagesRequestQueryValidatorTests
             };
-            var validator = new ListMessagesRequestQueryValidator();
+            var wrapper = new Mock<IRegexDateTimeWrapper>();
+            wrapper.Setup(p => p.IsMatch(It.IsAny<string>())).Returns(expected);
```
마찬가지로 여기서 `expected` 값은 이 `Validator.Validate(...)` 메소드의 결과값에 대한 예상치이지 이 목킹 결과값에 대한 예상치가 아니므로 이렇게 작성하면 테스트를 무조건 통과하는 의미없는 테스트 코드가 됩니다.

justinyoo commented on test/NhnToast.Sms.Tests/Validators/SendMessagesRequestBodyValidatorTests.cs:
```diff
@@ -48,7 +50,9 @@ public class SendMessagesRequestBodyValidatorTests
             };
-            var validator = new SendMessagesRequestBodyValidator();
+            var wrapper = new Mock<IRegexDateTimeWrapper>();
+            wrapper.Setup(p => p.IsMatch(It.IsAny<string>())).Returns(expected);
```
마찬가지로 여기서 `expected` 값은 이 `Validator.Validate(...)` 메소드의 결과값에 대한 예상치이지 이 목킹 결과값에 대한 예상치가 아니므로 이렇게 작성하면 테스트를 무조건 통과하는 의미없는 테스트 코드가 됩니다.

justinyoo reviewed (changes requested):
코멘트 남겼습니다!

MultiDCP committed:
Edit Field Convention of Validators

Edit Field Convention of Validators and Simplify if block in Wrapper

hsh108868 committed:
Modify test method

hsh108868 committed:
Merge branch 'main' of https://github.com/hsh108868/nhn-toast-notification-service-custom-connector

MultiDCP committed:
Merge remote-tracking branch 'upstream/main'

hsh108868 committed:
Merge pull request #7 from MultiDCP/main

justinyoo commented on src/nt-sms/Validators/RegexDateTimeWrapper.cs:
```diff
@@ -0,0 +1,34 @@
+        /// <inheritdoc />
+        public bool IsMatch(string date)
+        {
+            if (date == null)
```
혹시 이것도 `date.IsNullOrWhiteSpace()` 로 바꿀 수 있을 겁니다

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessageStatusRequestQueryValidatorTests.cs:
```diff
@@ -35,7 +36,10 @@ public void Given_Values_When_Validate_Invoked_Then_It_Should_Return_Result(stri
+            wrapper.Setup(p => p.IsMatch(It.IsAny<string>())).Returns(true);
+            var validator = new ListMessageStatusRequestQueryValidator(wrapper.Object);
 
             var result = validator.Validate(queries);
```
이 테스트의 목적을 한 번만 더 생각해 볼까요?

이 테스트는 `ListMessageStatusRequestQueryValidator.Validate()` 메소드가 결과값을 반환하는지를 테스트 하려는 거예요. 즉 `result.IsValid` 값이 `true` 인지 `false`인지만 보려는 겁니다. 이 때 정규식이 검사하려는 부분은 날짜/시간 형식인데, 이 테스트에서는 이 검사를 "했다 치고" 넘어가는 거죠. 실제 날짜 포맷에 상관 없이 무조건 날짜/시간 형식이 맞다(`true`) 혹은 틀리다(`false`)라고 가정한 후에 다른 것들만 체크하는 거예요.

* 이 날짜/시간 형식이 틀리다(`false`)라고 가정한다면 `result.IsValid` 값은 무조건 `false`가 되겠죠?
* 이 날짜/시간 형식이 맞다(`true`)라고 가정한다면 다른 유효성 검사값들 결과에 따라 `result.IsValid` 값이 결정되겠죠?

이런 가정 아래에서 `DataRow(...)`에 정의해 놓은 값들이 맞는 지도 확인해 보셔야 합니다.

---

만약에 `Mock<IRegexDateTimeWrapper>`를 사용하지 않고 실제 인스턴스인 `RegexDateTimeWrapper`를 사용한다면 어떨까요? 이 때에는 `DataRow(...)`에 정의해 놓은 값들이 맞나요?

justinyoo reviewed (changes requested):
거의 다 왔습니다. 조금만 힘 내 주세요!

MultiDCP committed:
Improve Date Null Checking in Wrapper

Improve Date Null Checking in RegexDateTimeWrapper by using IsNullOrWhiteSpace

MultiDCP committed:
Merge remote-tracking branch 'upstream/main'

hsh108868 committed:
Modify testcase

hsh108868 committed:
Merge pull request #8 from MultiDCP/main

수정사항 반영 - 널체크

MultiDCP committed:
Inject Dependency of RegexDateTimeWrapper

Add dependency injection by registering the IRegexDateTimeWrapper service with RegexDateTimeWrapper as Singleton

hsh108868 committed:
Merge pull request #9 from MultiDCP/main

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessageStatusRequestQueryValidatorTests.cs:
```diff
@@ -62,7 +65,11 @@ public void Given_InvalidValues_When_Validate_Invoked_Then_It_Should_Throw_Excep
             };
-            var validator = new ListMessageStatusRequestQueryValidator();
+
+            var wrapper = new Mock<IRegexDateTimeWrapper>();
+            wrapper.Setup(p => p.IsMatch(It.IsAny<string>())).Returns(false);
+            var validator = new ListMessageStatusRequestQueryValidator(wrapper.Object);
```
그렇다면 여기도 목킹하는 대신 실제 인스턴스를 넣어 보세요

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessageStatusRequestQueryValidatorTests.cs:
```diff
@@ -84,7 +91,10 @@ public async Task Given_ValidValues_When_Validate_Invoked_Then_It_Should_Return_
             };
-            var validator = new ListMessageStatusRequestQueryValidator();
+
+            var wrapper = new Mock<IRegexDateTimeWrapper>();
+            wrapper.Setup(p => p.IsMatch(It.IsAny<string>())).Returns(true);
+            var validator = new ListMessageStatusRequestQueryValidator(wrapper.Object);
```
그렇다면 여기도 목킹하는 대신 실제 인스턴스를 넣어 보세요

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessagesRequestQueryValidatorTests.cs:
```diff
@@ -105,7 +109,10 @@ public class ListMessagesRequestQueryValidatorTests
             };
-            var validator = new ListMessagesRequestQueryValidator();
+
+            var wrapper = new Mock<IRegexDateTimeWrapper>();
+            wrapper.Setup(p => p.IsMatch(It.IsAny<string>())).Returns(false);
+            var validator = new ListMessagesRequestQueryValidator(wrapper.Object);
```
그렇다면 여기도 목킹하는 대신 실제 인스턴스를 넣어 보세요

justinyoo commented on test/NhnToast.Sms.Tests/Validators/ListMessagesRequestQueryValidatorTests.cs:
```diff
@@ -141,7 +148,10 @@ public class ListMessagesRequestQueryValidatorTests
             };
-            var validator = new ListMessagesRequestQueryValidator();
+
+            var wrapper = new Mock<IRegexDateTimeWrapper>();
+            wrapper.Setup(p => p.IsMatch(It.IsAny<string>())).Returns(true);
+            var validator = new ListMessagesRequestQueryValidator(wrapper.Object);
```
그렇다면 여기도 목킹하는 대신 실제 인스턴스를 넣어 보세요

justinyoo commented on test/NhnToast.Sms.Tests/Validators/SendMessagesRequestBodyValidatorTests.cs:
```diff
@@ -88,7 +91,10 @@ public class SendMessagesRequestBodyValidatorTests
             };
-            var validator = new SendMessagesRequestBodyValidator();
+
+            var wrapper = new Mock<IRegexDateTimeWrapper>();
+            wrapper.Setup(p => p.IsMatch(It.IsAny<string>())).Returns(false);
+            var validator = new SendMessagesRequestBodyValidator(wrapper.Object);
```
그렇다면 여기도 목킹하는 대신 실제 인스턴스를 넣어 보세요

justinyoo commented on test/NhnToast.Sms.Tests/Validators/SendMessagesRequestBodyValidatorTests.cs:
```diff
@@ -126,7 +132,10 @@ public class SendMessagesRequestBodyValidatorTests
             };
-            var validator = new SendMessagesRequestBodyValidator();
+
+            var wrapper = new Mock<IRegexDateTimeWrapper>();
+            wrapper.Setup(p => p.IsMatch(It.IsAny<string>())).Returns(true);
+            var validator = new SendMessagesRequestBodyValidator(wrapper.Object);
```
그렇다면 여기도 목킹하는 대신 실제 인스턴스를 넣어 보세요

justinyoo reviewed (changes requested):
아마도 마지막 리뷰 코멘트가 될 것 같습니다. 이것만 확인해 보세욥!

hsh108868 committed:
Modify tests validator

justinyoo reviewed (approved):
수고하셨습니다!
